README.md
# Short Text [![Build Status][ci-img]][ci]

<img align="right" width="135" height="95" src="http://postcss.github.io/postcss/logo-leftp.png" title="Philosopher’s stone, logo of PostCSS">

[Short Text] is a [PostCSS] plugin that lets you use a shorthand `text` property in CSS.

This `text` property can map to several text-related properties, including `color`, `font-style`, `font-variant`, `font-weight`, `font-stretch`, `text-decoration`, `text-align`, `text-rendering`, `text-transform`, `white-space`, `font-size`, `line-height`, `letter-spacing`, and `word-spacing`.

```css
/* before */

section {
    text: bold center uppercase dimgrey 1.25em 1.5 .05em;
}

article {
    text: 1.25em * .05em;
}

/* after */

section {
    font-weight: bold;
    text-align: center;
    text-transform: uppercase;
    color: dimgrey;
    font-size: 1.25em;
    line-height: 1.5;
    letter-spacing: .05em;
}

article {
    font-size: 1.25em;
    letter-spacing: .05em;
}
```

Properties are matched into thematic groups, which can be written out of order. However, once a property matches a group, any other matching properties for that group must then come before properties of another group. Within a group, asterisks indicate that an individual property should be skipped. The groups are:

- `color`
- `font-style`, `font-variant`, `font-weight`, `font-stretch`
- `text-decoration`, `text-align`, `text-rendering`, `text-transform`
- `white-space`
- `font-size`, `line-height`, `letter-spacing`, `word-spacing`

## Usage

Follow these steps to use [Short Text].

Add [Short Text] to your build tool:

```bash
npm install postcss-short-text --save-dev
```

#### Node

```js
require('postcss-short-text')({ /* options */ }).process(YOUR_CSS);
```

#### PostCSS

Add [PostCSS] to your build tool:

```bash
npm install postcss --save-dev
```

Load [Short Text] as a PostCSS plugin:

```js
postcss([
    require('postcss-short-text')({ /* options */ })
]);
```

#### Gulp

Add [Gulp PostCSS] to your build tool:

```bash
npm install gulp-postcss --save-dev
```

Enable [Short Text] within your Gulpfile:

```js
var postcss = require('gulp-postcss');

gulp.task('css', function () {
    return gulp.src('./css/src/*.css').pipe(
        postcss([
            require('postcss-short-text')({ /* options */ })
        ])
    ).pipe(
        gulp.dest('./css')
    );
});
```

#### Grunt

Add [Grunt PostCSS] to your build tool:

```bash
npm install grunt-postcss --save-dev
```

Enable [Short Text] within your Gruntfile:

```js
grunt.loadNpmTasks('grunt-postcss');

grunt.initConfig({
    postcss: {
        options: {
            processors: [
                require('postcss-short-text')({ /* options */ })
            ]
        },
        dist: {
            src: 'css/*.css'
        }
    }
});
```

## Options

#### `prefix`

Type: `String`  
Default: `null`

Specifies a prefix to be surrounded by dashes before the declaration (e.g. `-x-text`).

[ci]: https://travis-ci.org/jonathantneal/postcss-short-text
[ci-img]: https://travis-ci.org/jonathantneal/postcss-short-text.svg
[Gulp PostCSS]: https://github.com/postcss/gulp-postcss
[Grunt PostCSS]: https://github.com/nDmitry/grunt-postcss
[PostCSS]: https://github.com/postcss/postcss
[Short Text]: https://github.com/jonathantneal/postcss-short-text


CHANGELOG.md
## 2.0.0 (2015-09-22)

Added: Transform properties by groups
Removed: Unused JavaScript file
Updated: Documentation

## 1.0.0 (2015-09-21)

Added: Initial version


test/test.js
var path    = require('path');
var postcss = require('postcss');
var expect  = require('chai').expect;
var fs      = require('fs');

var plugin = require('../');

function test(name, opts, done) {
	var fixtureDir = './test/fixtures/';
	var baseName   = name.split(':')[0];
	var testName   = name.split(':').join('.');
	var inputPath  = path.resolve(fixtureDir + baseName + '.css');
	var expectPath = path.resolve(fixtureDir + testName + '.expect.css');

	var inputCSS  = fs.readFileSync(inputPath, 'utf8');
	var expectCSS = fs.readFileSync(expectPath, 'utf8');

	postcss([plugin(opts)]).process(inputCSS, {
		from: inputPath
	}).then(function (result) {
		var actualCSS = result.css;

		expect(actualCSS).to.eql(expectCSS);
		expect(result.warnings()).to.be.empty;

		done();
	}).catch(function (error) {
		done(error);
	});
}

describe('postcss-short-text', function () {
	it('supports standard property', function (done) {
		test('basic', {}, done);
	});

	it('ignores standard property with prefix', function (done) {
		test('basic:w-prefix', { prefix: 'x' }, done);
	});

	it('ignores prefixed property', function (done) {
		test('prefixed', {}, done);
	});

	it('supports prefixed property with prefix', function (done) {
		test('prefixed:w-prefix', { prefix: 'x' }, done);
	});
});


css-get-value.js
module.exports = {
	getValueType: function getValueType(value) {
		var type = module.exports.type;

		for (var name in type) {
			if (type[name].test(value)) return name;
		}

		return 'keyword';
	},
	matchValueType: function matchValueType(value) {
		var index = -1;
		var types = Array.prototype.slice.call(arguments, 1);
		var test = module.exports.getValueType;

		while (type = types[++index]) {
			if (test(value) === type) {
				return true;
			}
		}

		return false;
	},
	type: {
		'border-style-keyword': /^(dashed|dotted|double|groove|inset|outset|ridge|solid)$/,
		'calc-function': /^calc\(.*\)$/,
		'color-hex': /^#([0-9A-f]{3}|[0-9A-f]{6})$/,
		'color-hsl-function': /^hsla?\(.+\)$/,
		'color-keyword': /^(aliceblue|antiquewhite|aqua|aquamarine|azure|beige|bisque|black|blanchedalmond|blue|blueviolet|brown|burlywood|cadetblue|chartreuse|chocolate|coral|cornflowerblue|cornsilk|crimson|currentColor|cyan|darkblue|darkcyan|darkgoldenrod|darkgray|darkgreen|darkgrey|darkkhaki|darkmagenta|darkolivegreen|darkorange|darkorchid|darkred|darksalmon|darkseagreen|darkslateblue|darkslategray|darkslategrey|darkturquoise|darkviolet|deeppink|deepskyblue|dimgray|dimgrey|dodgerblue|firebrick|floralwhite|forestgreen|fuchsia|gainsboro|ghostwhite|gold|goldenrod|gray|green|greenyellow|grey|honeydew|hotpink|indianred|indigo|ivory|khaki|lavender|lavenderblush|lawngreen|lemonchiffon|lightblue|lightcoral|lightcyan|lightgoldenrodyellow|lightgray|lightgreen|lightgrey|lightpink|lightsalmon|lightseagreen|lightskyblue|lightslategray|lightslategrey|lightsteelblue|lightyellow|lime|limegreen|linen|magenta|maroon|mediumaquamarine|mediumblue|mediumorchid|mediumpurple|mediumseagreen|mediumslateblue|mediumspringgreen|mediumturquoise|mediumvioletred|midnightblue|mintcream|mistyrose|moccasin|navajowhite|navy|oldlace|olive|olivedrab|orange|orangered|orchid|palegoldenrod|palegreen|paleturquoise|palevioletred|papayawhip|peachpuff|peru|pink|plum|powderblue|purple|rebeccapurple|red|rosybrown|royalblue|saddlebrown|salmon|sandybrown|seagreen|seashell|sienna|silver|skyblue|slateblue|slategray|slategrey|snow|springgreen|steelblue|tan|teal|thistle|tomato|transparent|turquoise|violet|wheat|white|whitesmoke|yellow|yellowgreen)$/i,
		'color-rgb-function': /^(hsl|rgb)a?\(.+\)$/,
		'font-style-keyword': /^(italic|oblique)$/,
		'font-weight-keyword': /^(100|200|300|400|500|600|700|800|900|black|bold|book|demibold|extrabold|extralight|heavy|light|medium|regular|roman|semibold|thin|ultrabold|ultralight)$/,
		'font-weight-relative-keyword': /^(bolder|lighter)$/,
		'global-keyword': /^(inherit|initial|unset)$/,
		'length': /^([-+]?0|[-+]?[0-9]*\.?[0-9]+(ch|cm|em|ex|in|mm|pc|pt|px|rem|vh|vmax|vmin|vw))$/,
		'length-absolute-size-keyword': /^(xx-small|x-small|small|medium|large|x-large|xx-large)$/,
		'length-keyword': /^(available|border-box|content-box|fit-content|max-content|min-content)$/,
		'length-percentage': /[-+]?[0-9]*\.?[0-9]+%/,
		'length-relative-size-keyword': /^(larger|smaller)$/,
		'position-keyword': /^(absolute|fixed|relative|static|sticky)$/,
		'var-function': /^var\(.*\)$/
	}
};


index.js
var assign  = require('object-assign');
var postcss = require('postcss');
var matches = {
	'color': /^(\*|aliceblue|antiquewhite|aqua|aquamarine|azure|beige|bisque|black|blanchedalmond|blue|blueviolet|brown|burlywood|cadetblue|chartreuse|chocolate|coral|cornflowerblue|cornsilk|crimson|currentColor|cyan|darkblue|darkcyan|darkgoldenrod|darkgray|darkgreen|darkgrey|darkkhaki|darkmagenta|darkolivegreen|darkorange|darkorchid|darkred|darksalmon|darkseagreen|darkslateblue|darkslategray|darkslategrey|darkturquoise|darkviolet|deeppink|deepskyblue|dimgray|dimgrey|dodgerblue|firebrick|floralwhite|forestgreen|fuchsia|gainsboro|ghostwhite|gold|goldenrod|gray|green|greenyellow|grey|honeydew|hotpink|indianred|indigo|inherit|ivory|khaki|lavender|lavenderblush|lawngreen|lemonchiffon|lightblue|lightcoral|lightcyan|lightgoldenrodyellow|lightgray|lightgreen|lightgrey|lightpink|lightsalmon|lightseagreen|lightskyblue|lightslategray|lightslategrey|lightsteelblue|lightyellow|lime|limegreen|linen|magenta|maroon|mediumaquamarine|mediumblue|mediumorchid|mediumpurple|mediumseagreen|mediumslateblue|mediumspringgreen|mediumturquoise|mediumvioletred|midnightblue|mintcream|mistyrose|moccasin|navajowhite|navy|oldlace|olive|olivedrab|orange|orangered|orchid|palegoldenrod|palegreen|paleturquoise|palevioletred|papayawhip|peachpuff|peru|pink|plum|powderblue|purple|rebeccapurple|red|rosybrown|royalblue|saddlebrown|salmon|sandybrown|seagreen|seashell|sienna|silver|skyblue|slateblue|slategray|slategrey|snow|springgreen|steelblue|tan|teal|thistle|tomato|transparent|turquoise|violet|wheat|white|whitesmoke|yellow|yellowgreen|#[0-9a-f]+|(hsl|rgb)a?\(.+\))$/i,
	'font-style': /^(\*|var\(.+\)|inherit|italic|normal|oblique)$/i,
	'font-variant': /^(\*|var\(.+\)|all-petite-caps|all-small-caps|inherit|none|normal|oldstyle-nums|ordinal|petite-caps|slashed-zero|small-caps|stacked-fractions|titling-caps|unicase)$/i,
	'font-weight': /^(\*|var\(.+\)|100|200|300|400|500|600|700|800|900|black|bold|bolder|book|demibold|extrabold|extralight|heavy|inherit|light|lighter|medium|normal|regular|roman|semibold|thin|ultrabold|ultralight)$/i,
	'font-stretch': /^(\*|var\(.+\)|condensed|expanded|extra-condensed|extra-expanded|inherit|normal|semi-condensed|semi-expanded|ultra-condensed|ultra-expanded)$/i,
	'text-decoration': /^(\*|var\(.+\)|blink|inherit|initial|line-through|none|overline|underline)$/i,
	'text-align': /^(\*|var\(.+\)|center|end|inherit|justify|left|match-parent|right|start|start-end)$/i,
	'text-rendering': /^(\*|var\(.+\)|auto|geometricPrecision|inherit|optimizeLegibility|optimizeSpeed)$/i,
	'text-transform': /^(\*|var\(.+\)|capitalize|full-width|inherit|lowercase|none|uppercase)$/i,
	'white-space': /^(\*|var\(.+\)|inherit|normal|nowrap|pre|pre-line|pre-wrap)$/i,
	'font-size': /^(\*|(calc|var)\(.+\)|inherit|initial|large|larger|medium|small|smaller|unset|x-large|x-small|xx-large|xx-small|0|[-+]?[0-9]*\.?[0-9]+(%|ch|cm|em|ex|in|mm|pc|pt|px|rem|vh|vmax|vmin|vw))$/i,
	'line-height': /^(\*|(calc|var)\(.+\)|inherit|initial|normal|unset|0|[-+]?[0-9]*\.?[0-9]+(%|ch|cm|em|ex|in|mm|pc|pt|px|rem|vh|vmax|vmin|vw)?)$/i,
	'letter-spacing': /^(\*|(calc|var)\(.+\)|inherit|initial|normal|unset|0|[-+]?[0-9]*\.?[0-9]+(%|ch|cm|em|ex|in|mm|pc|pt|px|rem|vh|vmax|vmin|vw))$/i,
	'word-spacing': /^(\*|(calc|var)\(.+\)|inherit|initial|normal|unset|0|[-+]?[0-9]*\.?[0-9]+(%|ch|cm|em|ex|in|mm|pc|pt|px|rem|vh|vmax|vmin|vw))$/i,
	'font-family': /^.+$/
};

module.exports = postcss.plugin('postcss-short-text', function (opts) {
	var name = opts && opts.prefix ? '-' + opts.prefix + '-text' : 'text';

	return function (css) {
		css.walkDecls(name, function (decl) {
			var props = assign({}, matches);
			var values = postcss.list.space(decl.value);
			var index = -1;
			var value;
			var prop;

			while (value = values[++index]) {
				for (prop in props) {
					if (props[prop].test(value)) {
						decl.cloneBefore({ prop: prop, value: value });

						delete props[prop];

						break;
					}
				}
			}

			decl.remove();
		});
	};
});
